File: hungry_critic_api/src/models/restaurant.ts
import { Allow, IsDefined, IsNumber } from 'class-validator';
import { Review } from './review';

export class Restaurant {
  @IsDefined()
  name: string;

  @Allow()
  address: string;

  @Allow({ each: true })
  cuisines: string[];
}

export class RestaurantDetails extends Restaurant {
  @Allow()
  id: string;

  @Allow()
  owner: string;

  @IsNumber({ maxDecimalPlaces: 0 })
  totalRatings: number;

  @IsNumber({ maxDecimalPlaces: 0 })
  totalReviews: number;

  @IsNumber()
  averageRating: number;

  bestReview?: Review;

  worstReview?: Review;
}

export class FilterCriteria {
  public minRating: number;
  public maxRating: number;

  constructor(query: any) {
    const rating = query['rating'];
    if (rating) {
      if (typeof rating === 'string') {
        this.initRating(rating);
      } else {
        rating.forEach((r: string) => this.initRating(r));
      }
    }

    this.minRating = this.minRating ?? 0;
    this.maxRating = this.maxRating ?? 5;
  }

  makeQuery() {
    const query: any = {};
    if (this.minRating) query.minRating = this.minRating;
    if (this.maxRating) query.maxRating = this.maxRating;
    return query;
  }

  private initRating(filter: String) {
    const parts = filter.split(':');
    switch (parts[0]) {
      case 'lte':
        this.maxRating = parseFloat(parts[1]);
        if(this.maxRating > 5) this.maxRating = 5;
        break;
      case 'gte':
        this.minRating = parseFloat(parts[1]);
        if(this.minRating < 0) this.minRating = 0;
        break;
    }
  }
}

File: hungry_critic/lib/models/restaurant.dart
import 'package:hungry_critic/models/review.dart';
import 'package:json_annotation/json_annotation.dart';

part 'restaurant.g.dart';

@JsonSerializable()
class Restaurant {
  Restaurant({
    required this.id,
    required this.owner,
    required this.name,
    this.address,
    this.cuisines = const [],
  })  : averageRating = 0,
        totalRatings = 0,
        totalReviews = 0;

  factory Restaurant.fromJson(Map<String, dynamic> json) => _$RestaurantFromJson(json);

  final String id;

  late String owner;

  final String name;

  final String? address;

  final List<String> cuisines;

  double averageRating;

  int totalRatings;

  int totalReviews;

  Review? bestReview;

  Review? worstReview;

  Map<String, dynamic> toJson() => _$RestaurantToJson(this);
}
